presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeConstantValuesUnderUnion.java

```diff
@@ -0,0 +1,131 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner.iterative.rule;
+
+import com.facebook.presto.Session;
+import com.facebook.presto.matching.Capture;
+import com.facebook.presto.matching.Captures;
+import com.facebook.presto.matching.Pattern;
+import com.facebook.presto.spi.plan.PlanNode;
+import com.facebook.presto.spi.plan.UnionNode;
+import com.facebook.presto.spi.plan.ValuesNode;
+import com.facebook.presto.spi.relation.RowExpression;
+import com.facebook.presto.spi.relation.VariableReferenceExpression;
+import com.facebook.presto.sql.planner.iterative.Rule;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import static com.facebook.presto.SystemSessionProperties.isOptimizeUnionOverValues;
+import static com.facebook.presto.matching.Capture.newCapture;
+import static com.facebook.presto.sql.planner.plan.Patterns.sources;
+import static com.facebook.presto.sql.planner.plan.Patterns.union;
+
+public class MergeConstantValuesUnderUnion
+        implements Rule<UnionNode>
+{
+    private static final Capture<List<PlanNode>> CHILDREN = newCapture();
+    private static final Pattern<UnionNode> PATTERN = union()
+            .with(sources().capturedAs(CHILDREN));
+
+    @Override
+    public Pattern<UnionNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public boolean isEnabled(Session session)
+    {
+        return isOptimizeUnionOverValues(session);
+    }
+
+    @Override
+    public Result apply(UnionNode node, Captures captures, Context context)
+    {
+        List<PlanNode> values = captures.get(CHILDREN);
+        //Return if not union over ValuesNode
+        if (!(values.stream().map(x -> context.getLookup().resolve(x)).allMatch(x -> x instanceof ValuesNode))) {
```

Comment:
We don't need all the children to be values nodes.  If any are values nodes, we can merge those and leave the other children as they were.

Reply:
@rschlussel As if now this optimizer replaces a Union node whose all children are valuesnode with a single values node.
What you are asking for is to subsume all valuesNode under union node ino the valuesnode.
I can :
1. Modify this optimizer to convert ot valuesnode or subsume in union depending upon whether all child are valuesnode
2. Implement a separate optimizer

Reply:
Good question.  I think option 1 is best.